util/topsql/reporter/reporter.go

```diff
@@ -563,14 +652,28 @@ func (tsr *RemoteTopSQLReporter) doReport(data reportData) {
 			}
 		}
 	})
-	ctx, cancel := context.WithTimeout(tsr.ctx, timeout)
-	start := time.Now()
-	err := tsr.client.Send(ctx, agentAddr, data)
-	if err != nil {
-		logutil.BgLogger().Warn("[top-sql] client failed to send data", zap.Error(err))
-		reportAllDurationFailedHistogram.Observe(time.Since(start).Seconds())
-	} else {
-		reportAllDurationSuccHistogram.Observe(time.Since(start).Seconds())
+
+	deadline := time.Now().Add(timeout)
+	for i := range tsr.dataSinks {
+		tsr.dataSinks[i].Send(data, deadline)
 	}
-	cancel()
+}
+
+var _ DataSinkRegHandle = &RemoteDataSinkRegHandle{}
+
+// RemoteDataSinkRegHandle is used to receive DataSink registrations.
+type RemoteDataSinkRegHandle struct {
+	registerCh chan DataSink
+}
+
+// Register implements DataSinkRegHandle interface.
+func (r *RemoteDataSinkRegHandle) Register(dataSink DataSink) (err error) {
+	defer func() {
+		if recover() != nil {
+			err = errors.New("registration channel is closed")
+		}
+	}()
+
+	r.registerCh <- dataSink
```

Comment:
It's not a good practice to write to a closed channel and use recover() to catch it. Could you find some other ways (or maybe you don't need to close the channel?)?